### ISSUE TITLE
Timestamp, Moderation Functions, Minimize Button 

### ISSUE THREAD

**Issue by warpunish** [Push-EDX/flarum-ext-chat#34]
Hello, 

We have been using your chat extension on our website for 4 months. Right now it works as good as it should be but it needs some improvements. We are getting nice traffic on our website and chat needs a delete message function because as you know, there are trolls everywhere. What it should do is delete a message if needed and update everyones chat frame as message deleted. Our users want timestamp for their messages. Right now when someone sends a message its unknown when it was sent. Another issue is chat minimization works perfect but it doesn't give a hint that when user clicks header bar it minimizes the frame. So what would be better is adding a minimize button to minimize chat frame. I know these are too much to ask and you might be busy. I am a ready to fork it to add these but i don't know which function does what. So if you are not willing to add these options, i am asking for a lead for what should i do to add these improvements.

**Comment by jtojnar** [comment]
* Weird, I thought I had opened a pull requests with timestamps in the title but it apparently is not the case. See https://github.com/Push-EDX/flarum-ext-chat/commit/e2b6a9096bc15aec9ef5a0f7f88bd9c00de48810 on general idea how to implement it.

* The header bar is described here https://github.com/Push-EDX/flarum-ext-chat/blob/63847ba0286142611816f7f1efda17c8b08b8be4/js/forum/src/components/ChatFrame.js#L267-L285

* The moderation will be the most complex, you will need to:
    * add a `pushedx.chat.delete` permission to `js/admin/src/main.js`
    * create a `DeleteChatController`, you can base it on `src/Api/Controllers/FetchChatController.php`
    * maybe create a `DeleteChatSerializer`?
    * create a `DeleteChat` command, see `src/Commands/PostChat.php` and a handler
    * add an API route and `canDeleteChat` attribute in `src/Listeners/AddChatApi.php`
    * add an interface; in the simplest case, you will want to add a delete button somewhere in the message wrapper
    * add translations

IDK, maybe I missed something, basically you need to go through every file and either edit it or clone it with a Delete in name.

**Comment by gpascualg** [comment]
@jtojnar Could you PR the timestamps? Or have you done any work towards moderation? I'm way to busy right now to do it :(

**Comment by jtojnar** [comment]
I am quite busy as well. If you want to use it, the commit is in [this branch](https://github.com/jtojnar/flarum-ext-chat/tree/time). Not sure why I did not open a PR; we have been using it on https://ostrov-tucnaku.cz/ for a while now so there should not be any serious issue.

**Comment by gpascualg** [comment]
No problem! You've done a lot of work already, I'll manually cherry-pick it later, thx!

**Comment by hifihedgehog** [comment]
I am on shared hosting without SSH so I use Bazaar exclusively. Is there a way I might be able to grab the moderation bits via Bazaar? Or should I simply be able to upload jtojnar's files over top gpascualg's without any trouble? Thank you everyone for your help over the last few days.

**Comment by jtojnar** [comment]
There are no moderation bits implemented. The only unmerged change are the [timestamp tooltips](https://github.com/jtojnar/flarum-ext-chat/tree/time) and [Telegram notifications hack](https://github.com/jtojnar/flarum-ext-chat/tree/other). Just copying the files over will not work since the same files have been modified. You will need to use git to merge the conflicts before copying the resulting sources to your server.

**Comment by hifihedgehog** [comment]
> Just copying the files over will not work since the same files have been modified. 

If I understand correctly, I can just download your fork this way, correct? I am not saying I would download from gpascualg's repo.
https://github.com/jtojnar/flarum-ext-chat

**Comment by jtojnar** [comment]
As GitHub says, “This branch is 6 commits behind Push-EDX:master”, the [`ot`](https://github.com/jtojnar/flarum-ext-chat/tree/ot) branch is probably up to date, but you might want to use `ot^` if you do not use Telegram.

**Comment by gpascualg** [comment]
@jtojnar Maybe you could do some PRs for that? Having some settings to toggle on/off telegram notifications and so on

**Comment by jtojnar** [comment]
I thought you were [going to cherry-pick](https://github.com/Push-EDX/flarum-ext-chat/issues/34#issuecomment-358770624) the [timestamp commit](https://github.com/jtojnar/flarum-ext-chat/commit/e2b6a9096bc15aec9ef5a0f7f88bd9c00de48810). As for the telegram code, it is jut a hack; I was waiting for some kind of hook to Flarum [Notify](https://discuss.flarum.org/d/4345-notify-notifications-for-slack-hipchat-and-gitter-compatible-with-beta-6) extension.

**Comment by gpascualg** [comment]
Yeah.. completely forgot about it, I've been disconnected from flarum lately. I've just cherry-picked it.

Ok, sounds like a good idea to wait for Notify extension
